File: elixir/space-age/lib/space_age.ex
defmodule SpaceAge do
  @type planet ::
          :mercury
          | :venus
          | :earth
          | :mars
          | :jupiter
          | :saturn
          | :uranus
          | :neptune

  @earth_orbital_period 31_557_600

  @doc """
  Return the number of years a person that has lived for 'seconds' seconds is
  aged on 'planet'.
  """
  @spec age_on(planet, pos_integer) :: float
  def age_on(:mercury, seconds) do
    seconds / (@earth_orbital_period * 0.2408467)
  end

  def age_on(:venus, seconds) do
    seconds / (@earth_orbital_period * 0.61519726)
  end

  def age_on(:earth, seconds) do
    seconds / @earth_orbital_period
  end

  def age_on(:mars, seconds) do
    seconds / (@earth_orbital_period * 1.8808158)
  end

  def age_on(:jupiter, seconds) do
    seconds / (@earth_orbital_period * 11.862615)
  end

  def age_on(:saturn, seconds) do
    seconds / (@earth_orbital_period * 29.447498)
  end

  def age_on(:uranus, seconds) do
    seconds / (@earth_orbital_period * 84.016846)
  end

  def age_on(:neptune, seconds) do
    seconds / (@earth_orbital_period * 164.79132)
  end
end

File: go/space-age/space_age.go
package space

type Planet string

const earthOrbitalPeriod = 31557600 // in seconds

func Age(seconds float64, planet Planet) float64 {

	switch planet {
	case "Mercury":
		return seconds / (earthOrbitalPeriod * 0.2408467)
	case "Venus":
		return seconds / (earthOrbitalPeriod * 0.61519726)
	case "Earth":
		return seconds / earthOrbitalPeriod
	case "Mars":
		return seconds / (earthOrbitalPeriod * 1.8808158)
	case "Jupiter":
		return seconds / (earthOrbitalPeriod * 11.862615)
	case "Saturn":
		return seconds / (earthOrbitalPeriod * 29.447498)
	case "Uranus":
		return seconds / (earthOrbitalPeriod * 84.016846)
	case "Neptune":
		return seconds / (earthOrbitalPeriod * 164.79132)
	default:
		return 0
	}

}
